Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,403 @@
+* `buffer_size` - On input this points to the size of the `result_buffer` in `char_t` units. On output this points to the number of `char_t` units used from the `result_buffer` (including the null terminator). If `result_buffer` is `nullptr` the input value is ignored and only the minimum required size in `char_t` units is set on output.
+* `assembly_path` - Optional. Path to the component's assembly. Whether or not this is specified determines the behavior for locating the hostfxr library.
+  * If `nullptr`, `hostfxr` is located using the environment variable or global registration
+  * If specified, `hostfxr` is located as if the `assembly_path` is an application with `apphost`
+
+`nethost` library uses the `__stdcall` calling convention.
+
+
+## Improve API to run application and load components
+
+### Goals
+* All hosts should be able to use the new API (whether they will is a separate question as the old API has to be kept for backward compat reasons)
+* Hide implementation details as much as possible
+  * Make the API generally easier to use/understand
+  * Give the implementation more freedom
+  * Allow future improvements without breaking the API
+  * Consider explicitly documenting types of behaviors which nobody should take dependency on (specifically failure scenarios)
+* Extensible
+  * It should allow additional parameters to some of the operations without a need to add new exported APIs
+  * It should allow additional interactions with the host - for example modifying how the runtime is initialized via some new options, without a need for a completely new set of APIs
+
+### New scenarios
+The API should allow these scenarios:
+* Runtime properties
+  * Specify additional runtime properties from the native host
+  * Implement conflict resolution for runtime properties
+  * Inspect calculated runtime properties (the ones calculated by `hostfxr`/`hostpolicy`)
+
+It should be possible to ship with only some of these supported, then enable more scenarios later on.
+
+All the proposed APIs will be exports of the `hostfxr` library and will use the same calling convention and name mangling as existing `hostfxr` exports.
+
+### Initialize host context
+
+All the "initialize" functions will
+* Process the `.runtimeconfig.json`
+* Resolve framework references and find actual frameworks
+* Find the root framework (`Microsoft.NETCore.App`) and load the `hostpolicy` from it
+* The `hostpolicy` will then process all relevant `.deps.json` files and produce the list of assemblies, native search paths and other artifacts needed to initialize the runtime.
+
+The functions will NOT load the CoreCLR runtime. They just prepare everything to the point where it can be loaded.
+
+The functions return a handle to a new host context:
+* The handle must be closed via `hostfxr_close`.
+* The handle is not thread safe - the consumer should only call functions on it from one thread at a time.
```

Comment:
Different components might have different runtime configs and thus different framework requirements. One context would not be able to handle that.

See discussion in #5486 that it might be possible to use the returned runtime helper function to effectively load multiple components assuming the same frameworks. But that is an open question.

Reply:
@sdmaclea This could be made to only effectively block on the first initialization. Subsequent ones could be made non-blocking. Does that alleviate your concern?

Reply:
The design is that the host context handle is not thread safe, but once it's used to get the runtime helper (through `hostfxr_get_runtime_delegate`) that helper should probably be thread safe and reusable (if it makes sense). Ultimately that is up to the helper implementation.

The helper design is being discussed in #5486 .